### PR TITLE
Implement Error Boundary

### PR DESCRIPTION
**Pull request by shami-dev**
Issue: #69

## Summary 
This pull request introduces the implementation of the Remix ErrorBoundary component in the root (parent) directory. This is the last resort for catching JavaScript errors. More specific errors can be handled inside child components. 

- Added a Remix ErrorBoundary component to root.tsx. It uses a wrapper that allows not to break UI
- Changed some import types because of the changes in Remix version 2
- Created ErrorMessage reusable component with basic styling
- Added a Remix ErrorBoundary component to _index.tsx
- Created and added a custom 404 image for handling a "Page Not Found" case
- Moved titles to meta and added conditions 

## Details (reason and description of the changes)
Users are happier if they are not bombarded with cumbersome error messages. Developers are happier when important information within errors isn't exposed to the client side. 

### Bugs 🐛 
- Use "Error" as a name for a component
- Place {children} inside html p tag
- Color contrast failed (fixed by using https://dequeuniversity.com/color-contrast)

### Alternative implementations (what alternatives were considered and why were they disregarded?) (optional)
- Using a 'Splat' route to handle the "Page Not Found" case

## Screenshots/Recordings: 
https://github.com/social-plan-it/plan-it-social-web/assets/114109736/beb2d045-06bd-4be4-b00b-c4634bf75edd

Co-authored-by: @willbeaumont 
Co-authored-by: @eiffelwong1 
Co-authored-by: Ahmad Akhtar

### CODE REVIEW

File: app/root.tsx [hunk 1]
```diff
@@ -44,3 +63,53 @@ export default function App() {
...
+          <p>{error.data}</p>
+        </ErrorMessage>
+      </Shell>
+    );
+  } else if (error instanceof Error) {
+    return (
+      <Shell title="Something went wrong">
+        <ErrorMessage>
+          <h1>Error</h1>
+          <p>{error.message}</p>
```
Comment (tonydangblog) [inline app/root.tsx:102]: Just food for thought - For security reasons, we may not want to default to showing a raw error message when we go to production. For example, an error could be thrown by our db that includes info about which ORM and database we're using, which in turn gives a potential hacker more info to attack our web app.

File: app/routes/_index.tsx [hunk 2]
```diff
@@ -29,3 +29,29 @@ export default function Index() {
     </>
   );
 }
+
+export function ErrorBoundary() {
+  const error = useRouteError();
+
+  if (isRouteErrorResponse(error)) {
+    return (
+      <div>
```
Comment (tonydangblog) [inline app/routes/_index.tsx:38]: Should we use the same `ErrorMessage` component as above here as well?

**Review by tonydangblog — COMMENTED** [review]
Cool PR. I haven't done error handling in Remix before so I definitely learned a lot :)

I left some comments for possible improvements. lmk what you think!

File: app/root.tsx [hunk 3]
```diff
@@ -25,17 +35,26 @@ export async function loader({ request }: LoaderArgs) {
...
 
 export default function App() {
+  return (
+    <Shell>
+      <Outlet />
+    </Shell>
+  );
+}
+
+function Shell({ title, children }: { title?: string; children: React.ReactNode }) {
```
[the remark this answers is not in the document]
Comment (tonydangblog) [reply app/root.tsx:45]: Perhaps we should use the built-in Remix meta function to set page names instead?

File: app/root.tsx (hunk 1 above)
Comment (tonydangblog) [reply app/root.tsx:102]: It looks like Remix already sanitizes errors for us so we don't need to worry about this. Here is a reference link - https://remix.run/docs/en/main/guides/errors#error-sanitization

File: app/root.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrelandgraf) [reply app/root.tsx:45]: I think `meta` even receives an `errorResponse` object which we could check to dynamically create an error title!

File: app/root.tsx (hunk 1 above)
Comment (shami-dev) [reply app/root.tsx:102]: Great catch 👍

File: app/routes/_index.tsx (hunk 2 above)
Comment (shami-dev) [reply app/routes/_index.tsx:38]: We can use) I suggest deleting this ErrorBoundary component from index.tsx until we handle errors here differently. Hope it won't lead to performance issues.

File: app/root.tsx [hunk 4]
```diff
@@ -44,3 +63,53 @@ export default function App() {
...
+        </Shell>
+      );
+    }
+    return (
+      <Shell title="An Error Occurred">
+        <ErrorMessage>
+          <h1>
+            {error.status} {error.statusText}
+          </h1>
+          <p>{error.data}</p>
```
[the remark this answers is not in the document]
Comment (shami-dev) [reply app/root.tsx:93]: Interestingly, the Remix team does not use the optional chaining operator in their documentation. Which docs should we follow? :)

File: app/root.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (shami-dev) [reply app/root.tsx:45]: Sounds amazing. Let me try!

File: app/root.tsx [hunk 5]
```diff
@@ -1,19 +1,40 @@
...
 import { json } from '@remix-run/node';
 import stylesheet from './styles/tailwind.css';
 import { TopNav } from '~/components/layout/top-nav';
-import { Links, LiveReload, Meta, Outlet, Scripts, ScrollRestoration } from '@remix-run/react';
 import Footer from './components/layout/footer';
+import { ErrorMessage } from '~/components/ui/error-message';
 import { getCurrentUser } from '~/modules/session/session.server';
 import { db } from './modules/database/db.server';
 
+export const meta: MetaFunction = ({ error }: any) => {
```
Comment (eiffelwong1) [inline app/root.tsx:20]: :any can be removed as it's type is inferred by the MetaFunction type that's already there

**Review by eiffelwong1 — DISMISSED** [review]
left some comments, looking great overall.

File: app/root.tsx (hunk 5 above)
Comment (shami-dev) [reply app/root.tsx:20]: This is a workaround) Did it but TypeScript's complaining with "Property 'error' does not exist on type 'ServerRuntimeMetaArgs<unknown, Record<string, unknown>>'.ts(2339)". 
It also worked when I used the error argument without destructuring. Not sure if it's a good idea...

File: app/root.tsx [hunk 6]
```diff
@@ -1,19 +1,40 @@
...
+  const title = !error
+    ? 'Plan It Social'
+    : isRouteErrorResponse(error)
```
Comment (andrelandgraf) [inline app/root.tsx:23]: Should we use the `error.status` property to give the user some more concrete information/context about **what** error occurred?

For instance: 404 -> "Page not found"!

File: app/root.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrelandgraf) [reply app/root.tsx:103]:
```suggestion
{!!error.data && <p>{error.data}</p>}
```

**Review by andrelandgraf — APPROVED** [review]
LGTM! 🥳

File: app/root.tsx (hunk 6 above)
Comment (andrelandgraf) [reply app/root.tsx:23]: (No blocker, can be follow-up work!)

**Review by eiffelwong1 — APPROVED** [review]
new changes looks good to me

**Review by andrelandgraf — APPROVED** [review]

**Pull request merged**